oauth/lib/oauth2.cpp

```diff
@@ -0,0 +1,987 @@
+                // some services, such as Google (when using Native App type)
+                // require the client secret...
+                query.addQueryItem("client_secret", d->m_clientSecret);
+            }
+            break;
+        }
+        default: { // ResourceOwnerPasswordFlow
+            query.addQueryItem("grant_type", "password");
+            // username and password should be passed in customParameters.
+            break;
+        }
+    }
+
+    for (QVariantMap::const_iterator it = d->m_customParameters.constBegin();
+            it != d->m_customParameters.constEnd(); ++it) {
+        query.addQueryItem(it.key(), it.value().toString());
+    }
+
+    performAccessTokenRequest(query);
+}
+
+/*!
+  \brief Request a refreshed access token from the server
+
+  If the client application has received a refresh token from the server,
+  it can request a refreshed access token periodically without requiring
+  further user interaction.
+
+  This method allows the client application to trigger such a request with
+  the remote server, by providing the previously returned \a refreshToken.
+
+  The request will be directed to the \l{refreshEndpoint} if a value has
+  been specified for that property, otherwise the \l{tokenEndpoint} will
+  be used instead.
+
+  On success, the \l{receivedAccessToken()} signal will be emitted.
+  Otherwise, the \l{error} and \l{errorMessage} properties will be set,
+  and the \l{errorChanged()} and \l{errorMessageChanged()} signals will
+  be emitted.
+*/
+void SailfishOS::OAuth::OAuth2::refreshAccessToken(const QString &refreshToken)
+{
+    Q_D(OAuth2);
+    if (d->m_flowType != AuthorizationCodeFlow
+            || d->m_flowType != AuthorizationCodeWithPkceFlow) {
+        d->m_error = InvalidOperationError;
+        d->m_errorMessage = QStringLiteral("Can only refresh access tokens received via authorization_code flow");
+        emit errorChanged();
+        emit errorMessageChanged();
+        return;
+    }
+
+    if ((d->m_refreshEndpoint.isEmpty() && d->m_tokenEndpoint.isEmpty())
+            || d->m_clientId.isEmpty()) {
+        d->m_error = InvalidParametersError;
+        d->m_errorMessage = QStringLiteral("tokenEndpoint and clientId are required parameters for refresh_token flow");
+        emit errorChanged();
+        emit errorMessageChanged();
+        return;
+    }
+
+    if (d->m_flowType == AuthorizationCodeFlow
+            && d->m_clientSecret.isEmpty()) {
+        d->m_error = InvalidParametersError;
+        d->m_errorMessage = QStringLiteral("clientSecret is a required parameter for refresh_token flow without PKCE");
+        emit errorChanged();
+        emit errorMessageChanged();
+        return;
+    }
+
+    QUrlQuery query;
+    query.addQueryItem("grant_type", "refresh_token");
+    query.addQueryItem("refresh_token", refreshToken);
+    query.addQueryItem("client_id", d->m_clientId);
+    if (d->m_flowType == AuthorizationCodeFlow
```

Comment:
I think you can simplify the condition to check if the clientSecret is not empty. If the user sets it perhaps he want it to be used. Flow type you already checked above

Reply:
Good point, will remove those unnecessary conditions

Reply:
done